### PR TITLE
Groovy Interpreter for Apazhe Zeppelin [ZEPPELIN-2176]

### PR DESCRIPTION
**Pull request by dlukyanov**
### What is this PR for?
Groovy Interpreter

### What type of PR is it?
Feature 

### Todos
* [ Tests ] - Task
* [ Documentation ] - Task

### What is the Jira issue?
[ZEPPELIN-2176]

### How should this be tested?
Follow the groovy interpreter documentation samples

### Questions:
* Does the licenses files need update? YES
* Is there breaking changes for older versions? NO
* Does this needs documentation? YES

### CODE REVIEW

**Comment by dlukyanov** [comment]
@Leemoonsoo, I moved the repo under the user, used the latest master, now i have another error:
```
++ grep -e '"ref":' -e '"sha":'
++ tr '\n' ' '
+ COMMIT=
+ sleep 30
+ python ./travis_check.py dlukyanov
Traceback (most recent call last):
  File "./travis_check.py", line 36, in <module>
    commit = sys.argv[2]
IndexError: list index out of range
Build step 'Execute shell' marked build as failure
Putting comment on the pull request
Finished: FAILURE
```

**Comment by Leemoonsoo** [comment]
Jenkins build script didn't handle well when source branch name of pullrequest is `master`.
I just updated the jenkins build script to handle the name `master`. 
@dlukyanov Could you trigger jenkins again (close/open pr) and see if it checks correctly?

**Comment by dlukyanov** [comment]
@Leemoonsoo, now I see status from travis )
And going to solve strange build errors...
Thanks!

**Comment by Leemoonsoo** [comment]
The error 

```
[ERROR] error An unexpected error occurred: "https://registry.yarnpkg.com/binary-extensions/-/binary-extensions-1.8.0.tgz: Request failed \"502 Bad Gateway\"".
[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.3:yarn (yarn install) on project zeppelin-web: Failed to run task: 'yarn install --no-lockfile' failed. (error code 1) -> [Help 1]
[ERROR] 
```
looks like temporary network problem. You can always restart [the failed Job](https://travis-ci.org/dlukyanov/zeppelin/jobs/211224371) and see if it passing.

**Comment by dlukyanov** [comment]
@Leemoonsoo , is it possible to restart jenkins checks after travis succeed? 
I've never got travis success without restarting travis jobs manually.
https://travis-ci.org/dlukyanov/zeppelin/builds

**Comment by dlukyanov** [comment]
Hello All, Asking for help with Jenkins:
Travis succeed:  https://travis-ci.org/dlukyanov/zeppelin
But jenkins failed: https://builds.apache.org/job/zeppelin-pull-request/604/console
```[2017-03-21 06:01:28] Get build status ...
Build https://travis-ci.org/dlukyanov/zeppelin/builds/213281815
[1] OK              https://travis-ci.org/dlukyanov/zeppelin/jobs/213281816
[2] Not completed   https://travis-ci.org/dlukyanov/zeppelin/jobs/213281817
[3] OK              https://travis-ci.org/dlukyanov/zeppelin/jobs/213281818
[4] OK              https://travis-ci.org/dlukyanov/zeppelin/jobs/213281819
[5] OK              https://travis-ci.org/dlukyanov/zeppelin/jobs/213281820
[6] OK              https://travis-ci.org/dlukyanov/zeppelin/jobs/213281821
[7] OK              https://travis-ci.org/dlukyanov/zeppelin/jobs/213281822
[8] OK              https://travis-ci.org/dlukyanov/zeppelin/jobs/213281823
1 job(s) failed, 0 job(s) running/pending
+ RET_CODE=1
+ '[' 1 -eq 2 ']'
+ '[' 1 -eq 2 ']'
+ exit 1
```

**Comment by Leemoonsoo** [comment]
Looks like second test matrix succeeded after the manual restart.
In this case, you also need to trigger Jenkins again to check latest build status.
Simply close and reopen this PR will do that.

**Comment by Leemoonsoo** [comment]
I tested last commit and it works well.

However the variable can't be read from the other paragraph.
image

This behavior is different from other interpreters, like spark (scala), python, r.
Is it intended?

Also, can we place document under `/docs/interpreter/` instead of `/groovy/`, so website can show groovy documentation?

**Comment by dlukyanov** [comment]
@Leemoonsoo Interesting) I did not knew that... i'll take a look how it's implemented...
With doc - i'll move it

**Comment by dlukyanov** [comment]
@Leemoonsoo, Just committed changes.
Doc moved.
About shared vars - it will be a little different from scala/spark.
If you use groovy script variable (undeclared or annotated) it will behave like you mentioned

```java
a="Hello world"
//or
@groovy.transform.Field String aa = "Hello world"
//or
g.put('aaa',"Hello world")
```

groovyscriptvars

**Comment by dlukyanov** [comment]
@Leemoonsoo , done. waiting for travis.

**Comment by Leemoonsoo** [comment]
Thanks @dlukyanov for the contribution.

LGTM!

**Comment by Leemoonsoo** [comment]
Merge to master if no further comments.

**Comment by dlukyanov** [comment]
@AhyoungRyu, changes committed,waiting for travis

File: groovy/src/main/java/org/apache/zeppelin/groovy/GroovyInterpreter.java [hunk 1]
```diff
@@ -55,159 +55,172 @@
 
 /**
  * Groovy interpreter for Zeppelin.
+ * @author dlukyanov@ukr.net / dmitry lukyanov
```
Comment (AhyoungRyu) [inline groovy/src/main/java/org/apache/zeppelin/groovy/GroovyInterpreter.java]: Here also

File: groovy/src/main/java/org/apache/zeppelin/groovy/GObject.java [hunk 2]
```diff
@@ -41,308 +43,328 @@
 
 /**
  * Groovy interpreter for Zeppelin.
+ * @author dlukyanov@ukr.net / dmitry lukyanov
```
Comment (AhyoungRyu) [inline groovy/src/main/java/org/apache/zeppelin/groovy/GObject.java]: @dlukyanov Zeppelin generally don't write author name in the source code. Even if it does, all ppl still can track and see what you contribute to Zeppelin :)

File: groovy/src/main/java/org/apache/zeppelin/groovy/GroovyInterpreter.java [hunk 3]
```diff
@@ -55,159 +55,172 @@
...
-        log.info("groovy classes classpath: "+classes);
-        if(classes!=null && classes.length()>0 ){
-	        File fClasses = new File(classes);
-	        if(!fClasses.exists())fClasses.mkdirs();
-	        shell.getClassLoader().addClasspath(classes);
+      }
+
+      bindings.clear();
+      InterpreterResult result = new InterpreterResult(Code.SUCCESS, out.toString());
+      //log.info("RESULT: "+result);
```
Comment (AhyoungRyu) [inline groovy/src/main/java/org/apache/zeppelin/groovy/GroovyInterpreter.java]: Could you remove this commented `log.info` if it's not necessary?

**Comment by AhyoungRyu** [comment]
Thanks for addressing my comments! I left some minor comments again about "author name" in the source code and removing unnecessary `log.info`. Except these two things, LGTM 👍

File: groovy/src/main/java/org/apache/zeppelin/groovy/GroovyInterpreter.java (hunk 1 above)
Comment (dlukyanov) [inline groovy/src/main/java/org/apache/zeppelin/groovy/GroovyInterpreter.java]: code autoformat.. but it's not allowed to keep the `created by` ? any footprint? )))

Comment (AhyoungRyu) [inline groovy/src/main/java/org/apache/zeppelin/groovy/GroovyInterpreter.java]: @dlukyanov I see. Yeah so sadly Zeppelin doesn't encourage the footprint 😭 (see here: https://github.com/apache/zeppelin/commit/d64920d20b1d564faea431541a80dc876d35bd94)

**Comment by dlukyanov** [comment]
@Leemoonsoo , @AhyoungRyu , any other comments?

**Comment by AhyoungRyu** [comment]
Looks good to me :)

**Comment by AhyoungRyu** [comment]
Merge to master if no further comments.

**Pull request closed without merging**